integrationtests/src/test/java/org/neo4j/procedure/UserFunctionIT.java

```diff
@@ -0,0 +1,914 @@
+                    .next().get( "result" );
+        }
+
+        @UserFunction
+        public long recursiveSum( @Name( "someValue" ) long order )
+        {
+            if ( order == 0L )
+            {
+                return 0L;
+            }
+            else
+            {
+                Long prev =
+                        (Long) db.execute( "RETURN org.neo4j.procedure.recursiveSum({order}) AS someVal",
+                                map( "order", order - 1 ) )
+                                .next().get( "someVal" );
+                return order + prev;
+            }
+        }
+
+        @UserFunction
+        public long genericArguments( @Name( "strings" ) List<List<String>> stringList, @Name( "longs" ) List<List<List<Long>>> longList )
+        {
+            return stringList.size() + longList.size();
+        }
+
+        @UserFunction
+        public long mapArgument( @Name( "map" ) Map<String,Object> map )
+        {
+            return map.size();
+        }
+
+        @UserFunction
+        public Node node( @Name( "id" ) long id )
+        {
+            return db.getNodeById( id );
+        }
+
+        @UserFunction
+        public double squareDouble( @Name( "someValue" )double value )
+        {
+            return value * value;
+        }
+
+        @UserFunction
+        public double avgNumberList( @Name( "someValue" ) List<Number> list )
+        {
+            return list.stream().reduce( ( l, r ) -> l.doubleValue() + r.doubleValue() ).orElse( 0.0d ).doubleValue() /
+                   list.size();
+        }
+
+        @UserFunction
+        public double avgDoubleList(  @Name( "someValue" ) List<Double> list )
+        {
+            return list.stream().reduce( ( l, r ) -> l + r ).orElse( 0.0d ) / list.size();
+        }
+
+        @UserFunction
+        public long squareLong( @Name( "someValue" ) long value )
+        {
+            return value * value;
+        }
+
+        @UserFunction
+        public long throwsExceptionInStream()
+        {
+            throw new RuntimeException( "Kaboom" );
+        }
+
+        @UserFunction
+        public long indexOutOfBounds()
+        {
+            int[] ints = {1, 2, 3};
+            return ints[4];
+        }
+
+        @UserFunction
+        public List<String> listCoolPeopleInDatabase()
+        {
+            return db.findNodes( label( "Person" ) )
+                    .map( node -> (String) node.getProperty( "name" ) )
+                    .stream()
+                    .collect( Collectors.toList() );
+        }
+
+        @UserFunction
+        public long logAround()
+        {
+            log.debug( "1" );
+            log.info( "2" );
+            log.warn( "3" );
+            log.error( "4" );
+            return 1337L;
+        }
+
+        @UserFunction
+        public Node  readOnlyTryingToWrite()
+        {
+            return db.createNode();
+        }
+
+        @UserFunction
+        public Node readOnlyCallingWriteFunction()
```

Comment:
Oh I see what you mean the test is somewhat confusing but not completely meaningless. It asserts that we cannot get updates through a backdoor by calling another function that performs the actual update. There is another test that checks so that we cannot perform updates by calling a procedure that perform the actual update.